tests/unit/acquisition/test_optimizer.py

```diff
@@ -141,11 +150,30 @@ def test_continuous_optimizer(
     expected_maximizer: list[list[float]],
     optimizer: AcquisitionOptimizer,
 ) -> None:
-
     maximizer = optimizer(search_space, _quadratic_sum(shift))
     npt.assert_allclose(maximizer, expected_maximizer, rtol=1e-3)
 
 
+@pytest.mark.parametrize("num_optimization_runs", [1, 10])
```

Comment:
You could probably do more in depth testing by mocking gpflow.optimizers.Scipy().minimize so that it behaves like you want: e.g. fail three times then succeed. If it's not too difficult then I think that's definitely worth doing.

Reply:
I tried this but couldnt get it to work. Do you have any examples of mocking something like that?

Reply:
The following should work (though you'll need to fix the num_recovery_runs = 0 bug mentioned below first):

```python
@pytest.mark.parametrize("num_failed_runs", range(4))
@pytest.mark.parametrize("num_recovery_runs", range(4))
def test_optimize_continuous_recovery_runs(num_failed_runs, num_recovery_runs) -> None:

    scipy_minimize = Scipy.minimize
    failed_runs = 0

    def mock_minimize(self, *args, **kwargs):
        nonlocal failed_runs
        result = scipy_minimize(self, *args, **kwargs)
        if failed_runs < num_failed_runs:
            failed_runs += 1
            result.success = False
        return result

    with unittest.mock.patch(
            'trieste.acquisition.optimizer.gpflow.optimizers.Scipy.minimize',
            mock_minimize
    ):
        optimizer = generate_continuous_optimizer(
            num_optimization_runs=1, num_recovery_runs=num_recovery_runs
        )
        if num_failed_runs > num_recovery_runs:
            with pytest.raises(FailedOptimizationError) as e:
                optimizer(Box([-1], [1]), _quadratic_sum([0.5]))
        else:
            optimizer(Box([-1], [1]), _quadratic_sum([0.5]))
```